uber/models/department.py

```diff
@@ -0,0 +1,428 @@
+    'job_required_role',
+    MagModel.metadata,
+    Column('job_id', UUID, ForeignKey('job.id')),
+    Column('dept_role_id', UUID, ForeignKey('dept_role.id')))
+
+
+class DeptChecklistItem(MagModel):
+    department_id = Column(UUID, ForeignKey('department.id'))
+    attendee_id = Column(UUID, ForeignKey('attendee.id'))
+    slug = Column(UnicodeText)
+    comments = Column(UnicodeText, default='')
+
+    __table_args__ = (UniqueConstraint('department_id', 'slug'),)
+
+
+class DeptRole(MagModel):
+    name = Column(UnicodeText)
+    description = Column(UnicodeText)
+    department_id = Column(UUID, ForeignKey('department.id'))
+
+    dept_memberships = relationship(
+        'DeptMembership',
+        backref='dept_roles',
+        cascade='save-update,merge,refresh-expire,expunge',
+        secondary='dept_membership_dept_role')
+
+    __table_args__ = (UniqueConstraint('name', 'department_id'),)
+
+    @hybrid_property
+    def dept_membership_count(self):
+        return len(self.dept_memberships)
+
+    @dept_membership_count.expression
+    def dept_membership_count(cls):
+        return func.count(cls.dept_memberships)
+
+    @property
+    def dept_memberships_ids(self):
+        return [str(d.id) for d in self.dept_memberships]
+
+
+class DeptMembership(MagModel):
+    is_dept_head = Column(Boolean, default=False)
+    is_poc = Column(Boolean, default=False)
+    is_checklist_admin = Column(Boolean, default=False)
+    attendee_id = Column(UUID, ForeignKey('attendee.id'))
+    department_id = Column(UUID, ForeignKey('department.id'))
+
+    __mapper_args__ = {'confirm_deleted_rows': False}
+    __table_args__ = (UniqueConstraint('attendee_id', 'department_id'),)
+
+    @hybrid_property
+    def has_role(self):
+        return self.has_implicit_role or self.has_dept_role
+
+    @has_role.expression
+    def has_role(cls):
+        return or_(cls.has_implicit_role, cls.has_dept_role)
+
+    @hybrid_property
+    def has_implicit_role(self):
+        return self.is_dept_head or self.is_poc or self.is_checklist_admin
+
+    @has_implicit_role.expression
+    def has_implicit_role(cls):
+        return or_(
+            cls.is_dept_head == True,
+            cls.is_poc == True,
+            cls.is_checklist_admin == True)  # noqa: E712
+
+    @hybrid_property
+    def has_dept_role(self):
+        return bool(self.dept_roles)
+
+    @has_dept_role.expression
+    def has_dept_role(cls):
+        return exists().select_from(dept_membership_dept_role) \
+            .where(cls.id == dept_membership_dept_role.c.dept_membership_id)
+
+
+class DeptMembershipRequest(MagModel):
+    attendee_id = Column(UUID, ForeignKey('attendee.id'))
+
+    # A NULL value for the department_id indicates the attendee is willing
+    # to volunteer for any department (they checked "Anything" for
+    # "Where do you want to help?").
+    department_id = Column(UUID, ForeignKey('department.id'), nullable=True)
+
+    departments = relationship(
```

Comment:
If someone checks "Anywhere" for "Where do you want to help?" then `departments` will be a list of every department.

Reply:
Yea! But in all other cases, it returns one department, right? Because a new department request is create for each department they want to be in?

Again, nothing comes to mind to make this nicer, and I don't think it's a huge deal. It just might cause some confusion later if you have to refer to department_request.departments but only get one department in most cases. Folks reading the code down the road may assume that a single department request handles all the departments someone has requested.

Reply:
Hmm... perhaps I need to rethink this particular relationship. We could make `department_id` non-nullable, so DeptMembershipRequest works just like a normal association table. Then we could track the "Willing to work anywhere" state using an additional column on the attendee table.

It would simplify some things (like this), but _possibly_ make some things a little trickier.

Thoughts?

Reply:
Hm, I kind of like that approach better, but you've been working on this for a long time already -- do you think it will be significantly easier to do it now vs later? If not, I'd recommend leaving it as-is for now and re-evaluating later. Once this gets into admins' and volunteers' hands, I expect we'll have some iterating to do, and I don't want to overthink it before that. 😅